Add --display-times option to `install`, `upgrade`, and `reinstall`

apjanke opened the pull request:
- [X] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [X] Have you successfully run `brew style` with your changes locally?
- [X] Have you successfully run `brew tests` with your changes locally?

-----

Would you consider a `--display-build-times` option for `brew install` and `brew upgrade`? This could be helpful to developers who are wrangling big installation trees.

I'm working on a big build with lots of dependencies (like 40 minutes total), and I'd like to see which of the dependencies are taking a lot of compilation time, so I know where to spend my efforts at making relocatable bottles, removing dependencies, or something else. I don't see a good way to do this with a custom `brew` command, since there aren't public API hooks into the installation process.

Adds output like this:

```
$ brew install --display-build-times -s arpack cowsay libpng                                                   display-build-times
==> Downloading https://github.com/opencollab/arpack-ng/archive/3.6.0.tar.gz
...
==> make test
==> make install
🍺  /usr/local/Cellar/libpng/1.6.34: 26 files, 1.2MB, built in 33 seconds
==> Build times
arpack               28.326 s
cowsay               3.063 s
libpng               33.318 s
```

ilovezfs commented:
A comment from left field: this looks similar to what will be needed for https://github.com/Homebrew/brew/issues/4057

apjanke commented:
Indeed it does. Though if we're going to both, might want to implement them differently to avoid cluttering up the main `install`/`update` code.

apjanke commented:
Stand by... a PR for a Caveats with a common message-collector design is incoming...

apjanke commented:
Here's a design for a generic-ish message collector, used for Caveats, and could be extended for build times: https://github.com/Homebrew/brew/pull/4361

MikeMcQuaid commented on Library/Homebrew/cmd/install.rb:
```diff
@@ -241,10 +244,15 @@ def install
+      build_times = []
       formulae.each do |f|
         Migrator.migrate_if_needed(f)
+        t0 = Time.now
```
Technically this won't be a "build time" but a "installation time" as it includes e.g. pouring, post-install.

apjanke replied:
Changed the terminology to "install times".

MikeMcQuaid commented on docs/Manpage.md:
```diff
@@ -273,6 +273,9 @@ With `--verbose` or `-v`, many commands print extra debugging information. Note
     If `--verbose` (or `-v`) is passed, print the verification and postinstall steps.
 
+    If `--display-build-times` is passed, build times for each formula are printed
+    at the end of the upgrade run.
```
Not an upgrade here.

apjanke replied:
Amended to fix.

ilovezfs commented:
not reinstall?

apjanke commented:
Added reinstall.

Changed all the names to use "install times" instead of "build times".

Will change this to use #4361 once that gets merged.

apjanke commented:
Go away stalebot; I'm still working on this.

apjanke commented:
I've rewritten this to use the message-collector mechanism from #4361.

Added unit test.

Also changed `--display-install-times` to a more generic `--display-times`, so if we end up having other operations that support displaying timings, they can just reuse the same option.

MikeMcQuaid commented on Library/Homebrew/formula_installer.rb:
```diff
@@ -221,6 +221,7 @@ def build_bottle_postinstall
   end
 
   def install
+    t0 = Time.now
```
`start_time`

MikeMcQuaid commented on Library/Homebrew/formula_installer.rb:
```diff
@@ -349,7 +350,8 @@ def install
 
     opoo "Nothing was installed to #{formula.prefix}" unless formula.installed?
-    Homebrew.messages.formula_installed(formula)
+    t1 = Time.now
```
`end_time`

MikeMcQuaid commented on Library/Homebrew/messages.rb:
```diff
@@ -28,4 +33,12 @@ def display_caveats
+  def display_install_times
+    return if @install_times.empty?
+    oh1 "Installation times"
+    @install_times.each do |t|
```
`install_times.each`; may as well use that attribute reader.

MikeMcQuaid commented on Library/Homebrew/messages.rb:
```diff
@@ -28,4 +33,12 @@ def display_caveats
   end
+
+  def display_install_times
+    return if @install_times.empty?
```
`if install_times`; may as well use that attribute reader.

MikeMcQuaid commented on Library/Homebrew/cmd/install.rb:
```diff
@@ -245,6 +248,7 @@ def install
       return if formulae.empty?
       Install.perform_preinstall_checks
 
+      install_times = []
```
Where is this array used?

MikeMcQuaid commented on Library/Homebrew/cmd/reinstall.rb:
```diff
@@ -11,6 +14,7 @@ module Homebrew
   def reinstall
     FormulaInstaller.prevent_build_flags unless DevelopmentTools.installed?
 
+    install_times = []
```
Where is this array used?

MikeMcQuaid commented on Library/Homebrew/cmd/upgrade.rb:
```diff
@@ -91,6 +94,7 @@ def upgrade
       end
     end
 
+    install_times = []
```
Where is this array used?

MikeMcQuaid commented on Library/Homebrew/messages.rb:
```diff
@@ -1,23 +1,28 @@
 
   def display_messages
     display_caveats
+    if ARGV.include?("--display-times")
```
`display_install_times if ARGV.include?("--display-times")`

apjanke commented:
Nits addressed!

MikeMcQuaid commented:
Thanks again @apjanke!

apjanke commented:
You're welcome! 🎉
